auth/roles-metadata.cc

```diff
@@ -46,59 +46,42 @@ constexpr std::string_view qualified_name("system_auth.roles");
 
 future<bool> default_role_row_satisfies(
         cql3::query_processor& qp,
-        std::function<bool(const cql3::untyped_result_set_row&)> p) {
+        std::function<bool(const cql3::untyped_result_set_row&)> p,
+        std::optional<std::string> rolename) {
     static const sstring query = format("SELECT * FROM {} WHERE {} = ?",
             meta::roles_table::qualified_name,
             meta::roles_table::role_col_name);
 
-    return do_with(std::move(p), [&qp](const auto& p) {
-        return qp.execute_internal(
-                query,
-                db::consistency_level::ONE,
-                {meta::DEFAULT_SUPERUSER_NAME},
-                cql3::query_processor::cache_internal::yes).then([&qp, &p](::shared_ptr<cql3::untyped_result_set> results) {
-            if (results->empty()) {
-                return qp.execute_internal(
-                        query,
-                        db::consistency_level::QUORUM,
-                        internal_distributed_query_state(),
-                        {meta::DEFAULT_SUPERUSER_NAME},
-                        cql3::query_processor::cache_internal::yes).then([&p](::shared_ptr<cql3::untyped_result_set> results) {
-                    if (results->empty()) {
-                        return make_ready_future<bool>(false);
-                    }
-
-                    return make_ready_future<bool>(p(results->one()));
-                });
-            }
-
-            return make_ready_future<bool>(p(results->one()));
-        });
-    });
+    for (auto c : { db::consistency_level::ONE, db::consistency_level::QUORUM }) {
+        auto results = co_await qp.execute_internal(query, db::consistency_level::ONE
+            , {rolename.value_or(std::string(meta::DEFAULT_SUPERUSER_NAME))}
+            , cql3::query_processor::cache_internal::yes
+            );
+        if (!results->empty()) {
+            co_return p(results->one());
+        }
+    }
+    co_return false;
```

Comment:
I can't see what changed due to concurrent coroutinization.

Reply:
I still can't see it without replaying the coroutinization in my mind.

Reply:
so yes, I prefer to see the coroutinization first.

Reply:
It is there. There is an initial patch added: "[roles-metadata: Coroutinuze some helpers](https://github.com/scylladb/scylladb/pull/12214/commits/dbf0f4947debb59ce779542f17ad48285df5ce68)".